packages/eslint-plugin/src/rules/no-useless-default-assignment.ts

```diff
@@ -0,0 +1,263 @@
+                const paramType = checker.getTypeOfSymbol(params[paramIndex]);
+                if (!canBeUndefined(paramType)) {
+                  reportUselessDefault(node, 'parameter');
+                }
+              }
+            }
+          }
+        }
+        return;
+      }
+
+      // Handle destructuring patterns
+      if (parent.type === AST_NODE_TYPES.Property) {
+        // This is a property in an object destructuring pattern
+        const objectPattern = parent.parent as TSESTree.ObjectPattern;
+
+        const sourceType = getSourceTypeForPattern(objectPattern);
+        if (!sourceType) {
+          return;
+        }
+
+        const propertyName = getPropertyName(parent.key);
+        if (!propertyName) {
+          return;
+        }
+
+        const propertyType = getPropertyType(sourceType, propertyName);
+        if (!propertyType) {
+          return;
+        }
+
+        if (!canBeUndefined(propertyType)) {
+          reportUselessDefault(node, 'property');
+        }
+      }
+    }
+
+    function getSourceTypeForPattern(
+      pattern: TSESTree.ArrayPattern | TSESTree.ObjectPattern,
+    ): ts.Type | null {
+      let currentNode: TSESTree.Node = pattern;
+
+      // Walk up through nested patterns
+      while (
+        currentNode.parent.type === AST_NODE_TYPES.AssignmentPattern ||
+        currentNode.parent.type === AST_NODE_TYPES.Property ||
+        currentNode.parent.type === AST_NODE_TYPES.ObjectPattern ||
+        currentNode.parent.type === AST_NODE_TYPES.ArrayPattern
+      ) {
+        currentNode = currentNode.parent;
+      }
+
+      const parent = currentNode.parent;
+
+      // Handle variable declarator
+      if (parent.type === AST_NODE_TYPES.VariableDeclarator && parent.init) {
+        const tsNode = services.esTreeNodeToTSNodeMap.get(parent.init);
+        return checker.getTypeAtLocation(tsNode);
+      }
+
+      // Handle function parameter
+      if (
+        parent.type === AST_NODE_TYPES.FunctionExpression ||
+        parent.type === AST_NODE_TYPES.ArrowFunctionExpression ||
+        parent.type === AST_NODE_TYPES.FunctionDeclaration
+      ) {
+        const paramIndex = parent.params.indexOf(
+          currentNode as TSESTree.Parameter,
+        );
+        const tsFunc = services.esTreeNodeToTSNodeMap.get(parent);
+        const signature = checker.getSignatureFromDeclaration(tsFunc);
+        if (!signature) {
+          return null;
+        }
+        const params = signature.getParameters();
+        return checker.getTypeOfSymbol(params[paramIndex]);
+      }
+
+      return null;
+    }
+
+    function getPropertyName(
+      key: TSESTree.Expression | TSESTree.PrivateIdentifier,
+    ): string | null {
+      switch (key.type) {
+        case AST_NODE_TYPES.Identifier:
+          return key.name;
+        case AST_NODE_TYPES.Literal:
+          return String(key.value);
+        default:
+          return null;
+      }
+    }
```

Comment:
[Non-Actionable] I'm surprised a utility doesn't already exist for this - but I couldn't find one.

* `getStaticValue`: doesn't handle identifiers
* `getStringIfConstant`: doesn't handle identifiers (it's a small wrapper around `getStaticStringValue`)
* `getStaticStringValue`: didn't handle identifiers properly

Not requesting changes, just noting - maybe I missed one?